contrib/rocksdb_logtool.py

```diff
@@ -0,0 +1,198 @@
+#! /usr/bin/env python3
+
+"""rocksdb_logtool.py
+
+Provides uploading/downloading FoundationDB XML log files to Joshua cluster.
+
+This is a temporary solution for Joshua tests with RocksDB as the storage server
+of FoundationDB. Since RocksDB is *not* deterministic, FoundationDB simulation
+failures in Joshua may not be reproducible locally. With this script, it is
+possible to upload the XML logs on the Joshua side and download them to local
+directories.
+"""
+
+import argparse
+import logging
+import os
+import os.path
+import re
+import pathlib
+import subprocess
+import tempfile
+
+import fdb
+import joshua.joshua_model as joshua
+
+from typing import List
+
+# Defined in SimulatedCluster.actor.cpp:SimulationConfig::setStorageEngine
+ROCKSDB_TRACEEVENT_STRING = ["RocksDBNonDeterminism", "ShardedRocksDBNonDeterminism"]
+
+# e.g. /var/joshua/ensembles/20230221-051349-xiaogesu-c9fc5b230dcd91cf
+ENSEMBLE_ID_REGEXP = re.compile(r"ensembles\/(?P<ensemble_id>[0-9A-Za-z\-_]+)$")
+
+# e.g. <Test TestUID="1ad90d42-824b-4693-aacf-53de3a6ccd27" Statistics="AAAA
+TEST_UID_REGEXP = re.compile(r"TestUID=\"(?P<uid>[0-9a-fA-F\-]+)\"")
+
+logger = logging.getLogger(__name__)
+
+
+def _execute_grep(string: str, paths: List[pathlib.Path]) -> bool:
+    command = ["grep", "-F", string] + [str(path) for path in paths]
+    result = subprocess.run(command, stdout=subprocess.DEVNULL)
+    return result.returncode == 0
+
+
+def _is_rocksdb_test(log_files: List[pathlib.Path]) -> bool:
+    for event_str in ROCKSDB_TRACEEVENT_STRING:
+        if _execute_grep(event_str, log_files):
+            return True
+    return False
+
+
+def _extract_ensemble_id(work_directory: str) -> str:
+    match = ENSEMBLE_ID_REGEXP.search(work_directory)
+    if not match:
+        return None
+    return match.groupdict()["ensemble_id"]
+
+
+def _get_log_subspace(ensemble_id: str, test_uid: str):
+    ensemble_space = joshua.dir_ensembles
+    rocksdb_log_space = ensemble_space.create_or_open(joshua.db, "rocksdb_logs")
+    return rocksdb_log_space[bytes(ensemble_id, "utf-8")][bytes(test_uid, "utf-8")]
+
+
+def _tar_xmls(xml_files: List[pathlib.Path], output_file_name: pathlib.Path):
+    command = ["tar", "-c", "-f", str(output_file_name), "--xz"] + [
+        str(xml_file) for xml_file in xml_files
+    ]
+    logger.debug(f"Execute tar: {command}")
+    subprocess.run(command, check=True, stdout=subprocess.DEVNULL)
+
+
+def _tar_extract(path_to_archive: pathlib.Path):
+    command = ["tar", "xf", str(path_to_archive)]
+    subprocess.run(command, check=True, stdout=subprocess.DEVNULL)
+
+
+def report_error(
```

Comment:
Can we add another argument to force uploading logs? This can be useful when debugging other non-deterministic issues.